File: commonfamily/src/main/java/com/koenig/commonModel/finance/BookkeepingEntry.java
package com.koenig.commonModel.finance;

import com.koenig.commonModel.Item;
import com.koenig.commonModel.Validator;

import java.nio.ByteBuffer;

/**
 * Created by Thomas on 21.10.2017.
 */

public class BookkeepingEntry extends Item {

    // userId --> costs
    public CostDistribution costDistribution;
    private int costs;
    private String category;
    private String subCategory;


    public BookkeepingEntry(String name, String category, String subCategory, int costs, CostDistribution costDistribution) {
        super(name);
        this.costs = costs;
        this.category = category;
        this.subCategory = subCategory;
        this.costDistribution = costDistribution;
    }

    public BookkeepingEntry(BookkeepingEntry entry) {
        super(entry.getName());
        this.costs = entry.costs;
        this.category = entry.category;
        this.subCategory = entry.subCategory;
        this.costDistribution = entry.costDistribution;
    }

    public BookkeepingEntry(ByteBuffer buffer) {
        super(buffer);
        category = byteToString(buffer);
        subCategory = byteToString(buffer);
        costs = buffer.getInt();
        costDistribution = new CostDistribution(buffer);
    }


    public int getCosts() {
        return costs;
    }

    public void setCosts(int costs) {
        this.costs = costs;
    }

    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public String getSubCategory() {
        return subCategory;
    }

    public void setSubCategory(String subCategory) {
        this.subCategory = subCategory;
    }

    public CostDistribution getCostDistribution() {
        return costDistribution;
    }

    public void setCostDistribution(CostDistribution costDistribution) {
        this.costDistribution = costDistribution;
    }

    public boolean costsValid() {
        return costDistribution.sumReal() == costs;
    }

    @Override
    public int getByteLength() {
        return super.getByteLength() + getStringLength(category) + getStringLength(subCategory) + 4 + costDistribution.getByteLength();
    }

    @Override
    public void writeBytes(ByteBuffer buffer) {
        super.writeBytes(buffer);
        buffer.put(stringToByte(category));
        buffer.put(stringToByte(subCategory));
        buffer.putInt(costs);
        costDistribution.writeBytes(buffer);
    }

    public boolean isValid() {
        // subcategory is allowed to be empty
        return Validator.isNotEmpty(name) && Validator.isNotEmpty(category) &&
                costDistribution.isValid() && costDistribution.sumReal() == costs;
    }

    @Override
    public String toString() {
        return "BookkeepingEntry{" +
                "costs=" + costs +
                ", name='" + name + '\'' +
                ", category='" + category + '\'' +
                ", subCategory='" + subCategory + '\'' +
                ", costDistribution=" + costDistribution +
                '}';
    }
}

File: commonfamily/src/main/java/com/koenig/commonModel/finance/BookkeepingEntry.kt
package com.koenig.commonModel.finance

import com.koenig.commonModel.Byteable
import com.koenig.commonModel.Item
import com.koenig.commonModel.Validator

import java.nio.ByteBuffer

/**
 * Created by Thomas on 21.10.2017.
 */

open class BookkeepingEntry : Item {

    // userId --> costs
    var costDistribution: CostDistribution
    // costs in cent
    var costs: Int = 0
    var category: String
    var subCategory: String

    override val byteLength: Int
        get() = super.byteLength + Byteable.Companion.getStringLength(category) + Byteable.Companion.getStringLength(subCategory) + 4 + costDistribution.byteLength

    open// subcategory is allowed to be empty
    val isValid: Boolean
        get() = Validator.isNotEmpty(name) && Validator.isNotEmpty(category) &&
                costDistribution.isValid && costDistribution.sumReal() == costs


    constructor(name: String, category: String, subCategory: String, costs: Int, costDistribution: CostDistribution) : super(name) {
        this.costs = costs
        this.category = category
        this.subCategory = subCategory
        this.costDistribution = costDistribution
    }

    constructor(entry: BookkeepingEntry) : super(entry.name) {
        this.costs = entry.costs
        this.category = entry.category
        this.subCategory = entry.subCategory
        this.costDistribution = entry.costDistribution
    }

    constructor(buffer: ByteBuffer) : super(buffer) {
        category = Byteable.Companion.byteToString(buffer)
        subCategory = Byteable.Companion.byteToString(buffer)
        costs = buffer.int
        costDistribution = CostDistribution(buffer)
    }

    fun costsValid(): Boolean {
        return costDistribution.sumReal() == costs
    }

    override fun writeBytes(buffer: ByteBuffer) {
        super.writeBytes(buffer)
        buffer.put(Byteable.Companion.stringToByte(category))
        buffer.put(Byteable.Companion.stringToByte(subCategory))
        buffer.putInt(costs)
        costDistribution.writeBytes(buffer)
    }

    override fun toString(): String {
        return "BookkeepingEntry{" +
                "costs=" + costs +
                ", name='" + name + '\''.toString() +
                ", category='" + category + '\''.toString() +
                ", subCategory='" + subCategory + '\''.toString() +
                ", costDistribution=" + costDistribution +
                '}'.toString()
    }
}
